Initial load of Node session mgmt post (#809)
oktadev/okta-blog: _source/_posts/2021-06-07-session-mgmt-node.md

---
layout: blog_post
title: "Everything You Ever Wanted to Know About Session Management in Node.js"
author: nickolas-fisher
by: contractor
communities: [javascript]
description: "A quick run-down of how to manage sessions in Node.js."
tags: [session, node]
tweets:
- "Need to learn how to manage sessions in Node? We've got you covered!"
- "Learn how to manage sessions in Node!"
- "Managing sessions in Node can be confusing. Check this out!"
image: blog/featured/okta-node-skew.jpg

> type: conversion

Reviewer: This doesn't seem like conversion since there's no reference to http://developer.okta.com/signup nor a cli.md include.

---

Session Management is a pretty scary term for a lot of developers.  Most of the mechanics of session management are abstracted away from developers, to the point where they don't properly learn about it until it's necessary.  Usually, this means a late night trying to figure out a vulnerability, a bug, or how to work with a new session management library.  Hopefully, I can remove some of the magic behind session management in NodeJs and give you a good fundamental place to start.  In this article, you will learn what sessions are, how to manage them in Node.js, and some details that should help you minimize bugs and vulnerabilities.

> ## Overview of Sessions

Reviewer: It seems the keywords are session and node. If that's what you're targeting, it seems like Node.js should be in the headings more.

## Session Management Risks

As an aside, I should note that it's generally better to use a well-established session management library than to roll your own.  You will see a few examples of that later in this article.  For now, let's take a look at some common session security concerns.

### Session Prediction

Each session has an associated session ID with it.  It is very important that this session ID is properly randomized, such that an attacker cannot simply guess a few options and bypass any security associated with the session ID.  Suppose your session IDs were sequential integers.  An attacker could log in, create a new session, and see their session ID is `12345`.  The attacker could then try to pass the session ID `12344` or `12343` to the server in an attempt to hijack a session from another user.

### Session Sniffing

In session sniffing, an attacker can use a sniffing application such as Wireshark or a proxy to capture network traffic between a client and server.  As you've learned, that traffic will ultimately contain a request with a session cookie in it.  This cookie will have the session ID which can then be hijacked.

> ### Man in the middle attacks

Reviewer: ```suggestion
### Man-in-the-Middle Attacks
```

### Regeneration of Session After Login

When a user first accesses your site, he/she can use an anonymous session.  This is a fairly common practice where you want to track a user's movements in your application but don't want to require them to log in.  For example, consider a shopping site where an anonymous user can have a shopping cart but can't check out until he/she logs in.  In this case, there is still a session ID provided to the user.  When the user does log in, you should regenerate the session ID to prevent session fixation attacks against your web application.

## Session Management in Node.js

Node.js has become wildly popular over the last 10 years.  With it, several frameworks and session management libraries have cropped up to support it.  Most frameworks use their own session management middleware.  For example, `express`, the most popular server framework for Node.js, has the accompanying `express-session` for session management.  Similarly, `koa` uses `koajs/session` for its session management.  

> For this article, I want to focus on `express-session`, as it is one of the most mature and widely used session management libraries available for Node.js.  For a full rundown of the `express-session` package, you can view the [readme here](https://github.com/expressjs/session#readme).  Below aresome of the highlights:

Reviewer: ```suggestion
For this article, I want to focus on `express-session`, as it is one of the most mature and widely used session management libraries available for Node.js.  For a full rundown of the `express-session` package, you can view the [readme here](https://github.com/expressjs/session#readme).  Below are some of the highlights:
```

    },
    callback: {
      path: "/authorization-code/callback",
    },
    loginCallback: {
      afterCallback: "/users/afterLogin",
    },
  },
});

app.use(oidc.router);

app.get("/users/afterLogin", ensureAuthenticated, (request, response, next) => {
  request.session.regenerate(function(err) {
    // will have a new session here
  })
});
```

In the above example, you are registering the `ExpressOIDC` middleware provided by Okta to handle the login.  After the successful login, you redirect the users to the `users/afterLogin` router, which then has access to the `request` object.  `Express-session` has attached the `session` object to the request for you and you can use the `session` API to call `regenerate`.  This will create a new session for the logged-in users.  

## Learn More

Session management is a topic that you could spend days researching and understanding.  As I noted, the session management package you will use in Node.js will largely depend on your stack and your server framework.  However, by becoming familiar with implementing safe and optimized session management in one framework, you can carry that knowledge to all other Node.js frameworks.  

> 

Reviewer: Blog post links (below) should be relative.

> - [Why JWTs Suck as Session Tokens ](https://developer.okta.com/blog/2017/08/17/why-jwts-suck-as-session-tokens)

Reviewer: ```suggestion
- [Why JWTs Suck as Session Tokens](https://developer.okta.com/blog/2017/08/17/why-jwts-suck-as-session-tokens)
```

> - [Build a Simple CRUD Application with Node and MySQL ](https://developer.okta.com/blog/2019/09/09/build-crud-app-node-mysql)

Reviewer: ```suggestion
- [Build a Simple CRUD Application with Node and MySQL](https://developer.okta.com/blog/2019/09/09/build-crud-app-node-mysql)
```

> - [Build A Simple Web App with Node and Postgres ](https://developer.okta.com/blog/2019/11/22/node-postgres-simple-webapp)

Reviewer: ```suggestion
- [Build A Simple Web App with Node and Postgres](https://developer.okta.com/blog/2019/11/22/node-postgres-simple-webapp)
```
